src/libcore/str/pattern.rs

```diff
@@ -0,0 +1,359 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![allow(missing_docs)]
+
+use prelude::*;
+use super::CharEq;
+
+// Pattern
+
+pub trait Pattern<'a>: Sized {
+    type Searcher: Searcher<'a>;
+    fn into_searcher(self, haystack: &'a str) -> Self::Searcher;
+
+    #[inline]
+    fn is_contained_in(self, haystack: &'a str) -> bool {
+        self.into_searcher(haystack).next_match().is_some()
+    }
+
+    #[inline]
+    fn match_starts_at(self, haystack: &'a str, idx: usize) -> bool {
+        let mut matcher = self.into_searcher(haystack);
```

Comment:
Hm, actually, it may not. E.g. asking if a match for `abab` starts at 2 in `abababab` should return false, since the matches are indices 0-4 and 4-8, or does the searching look for all matching substrings, even if they overlap?